### ISSUE TITLE
Ethernet support (freertos) broken

### ISSUE THREAD

**Issue by kikass13** [modm-io/modm#758]
Hello, 

@rleh has requested me to investigate what happened to the ethernet support of the freertos implementation in the latest updates of modm (hint, ethernet broke).

looking into the history of the aws (https://github.com/modm-io/modm/commits/develop/ext/aws) I tested the commits one by one ... 

 * commit 1375ff15f569a8f21fa3c24bf57c1e25e2803eac seems to work
 * commit b1ac837c9c2fe24f55b5f8dcf92507667944c11f seems to break it

Ill try looking into this but it seems that this "major update" f'd up a lot.

**Comment by kikass13** [comment]
It turns out, that the scheduler is not working anymore ... i mean it crashes after some iterations even without all the ethernet stuff in my example :/


### code

All I have is a blinking led task and even that is gone after some time with the uc crashing and restarting ... 


```cpp

#include <modm/board.hpp>
#include <modm/processing/rtos.hpp>

// overwrite freertos config with our local one
#include <FreeRTOSConfigLocal.h>

using namespace Board;

static void myled(void* parameter) {
    MODM_LOG_DEBUG << "  -- Led blink task run()" << modm::endl;
    Leds::setOutput();
    while (true) {
        vTaskDelay(1000 * MILLISECONDS);
        Leds::toggle();
    }
}

int main() {
    Board::initialize();
    Leds::setOutput();
    MODM_LOG_INFO << "\n\n============================\nHello world ... \n" << modm::endl;
    modm::delay(1s);

    xTaskCreate(myled, "ledBLinkyboob", (configMINIMAL_STACK_SIZE * 2), 0, configMAX_PRIORITIES + 1, 0);
    MODM_LOG_INFO << "Initializing scheduler" << modm::endl;
    modm::rtos::Scheduler::schedule();
    // we should never get here
    return 0;
}
```

sp whatever [this update](https://github.com/modm-io/modm/commit/b1ac837c9c2fe24f55b5f8dcf92507667944c11f#diff-60aa089ae37e535e3725a2516d4f74c79fa9193e5ebfccb7666224d8d87a5781)  has done, it has broken freertos completely

### Output

```
============================
Hello world ... 

As

============================
Hello world ... 

As

============================
Hello world ... 

As

============================
Hello world ... 

As
```

**Comment by rleh** [comment]
You could include [CrashCatcher](https://modm.io/reference/module/modm-platform-fault/) in your example and try to find the reason for the (hard)fault.

**Comment by salkinium** [comment]
Not sure if that's the issue, but the config is automatically included by freertos, and must not be included by the application.

```cpp
// overwrite freertos config with our local one
#include <FreeRTOSConfigLocal.h>
```

**Comment by kikass13** [comment]
> Not sure if that's the issue, but the config is automatically included by freertos, and must not be included by the application.
> 
> ```c++
> // overwrite freertos config with our local one
> #include <FreeRTOSConfigLocal.h>
> ```

It does not matter ... It's not working either way :)

**Comment by kikass13** [comment]
> You could include [CrashCatcher](https://modm.io/reference/module/modm-platform-fault/) in your example and try to find the reason for the (hard)fault.

I will do that tomorrow. 

Were these changes important at all? Seems like a reverse and everything will be okay

**Comment by salkinium** [comment]
> Were these changes important at all? Seems like a reverse and everything will be okay

I mean it's the official FreeRTOS release, I doubt they release something so broken. There's probably been a subtle API change that our port is now missing. I did test the release and our standard `modm:processing:rtos` examples work fine, however, I didn't test the Ethernet example due to lack of hardware. However, the TCP code is quite large and I don't really know how it works, so I wouldn't have an advantage over you debugging it either…

**Comment by kikass13** [comment]
The example I posted does not utilize Ethernet, so I doubt that any of the examples works :/

**Comment by rleh** [comment]
> I doubt that any of the examples works :/

I just tested [this FreeRTOS example](https://github.com/modm-io/modm/tree/develop/examples/nucleo_f429zi/freertos_static) on a Nucleo-F439: Works fine!

**Comment by salkinium** [comment]
Your example crashes on my G071 too, however, if I set `configMAX_PRIORITIES - 1` as a priority it works fine. The priority is inverted, ie. any task must have a *LOWER* priority than `configMAX_PRIORITIES`, since that is used by the scheduler to interrupt the tasks and perform the context switch.

```
============================
Hello world ...

Initializing scheduler
  -- Led blink task run()
```

For a lot of detail on this see: https://www.freertos.org/RTOS-Cortex-M3-M4.html

**Comment by salkinium** [comment]
Or maybe I'm wrong and it's the other way around? Like `configMAX_PRIORITIES + 1` causes an overflow and sets the priority to zero, which is the highest priority?

**Comment by rleh** [comment]
This line triggers the assertion: https://github.com/modm-ext/freertos-partial/blob/dd1c0b244cb50bbe24d746478b0bf3839f3052d3/FreeRTOS/Source/tasks.c#L918

**Comment by rleh** [comment]
Fix: #760

**Comment by salkinium** [comment]
No, I'm not convinced, I think there's something else going on as well.
For example the application works fine with this change in `profile=release`, but crashes with `profile=debug` somewhere way before ever getting to `xTaskCreate`.
In addition the assertion is called from Undefined_Handler, rather than from `configASSERT`.

**Comment by rleh** [comment]
> For example the application works fine with this change in `profile=release`, but crashes with `profile=debug` somewhere way before ever getting to `xTaskCreate`.

Which example application? For me both the new FreeRTOS blink and ethernet examples work fine on the Nucleo-F767ZI. in release and debug mode.

> In addition the assertion is called from Undefined_Handler, rather than from `configASSERT`.

Are you sure? My stack indicates that `Undefined_Handler` is called from somewhere (`??`) from `modm::IOStream::operator<<` from `modm_assert_report` from `prvInitialiseNewTask` from `xTaskCreate`, which does not look wrong to me.

**Comment by salkinium** [comment]
Eh, I was trying to debug the example posted above, but I cannot pin down the issues. If it works in the examples I'm fine with it.

**Comment by kikass13** [comment]
hmm thank,s that solves that problem (i dont really know where my +1 came from and why this worked before).

So i can now start debugging the real problem with the [ethernet example](https://github.com/modm-io/modm/tree/develop/examples/nucleo_f767zi/ethernet). Currently, the ethernet example does not do anything

```cpp
int
main()
{
	Board::initialize();
	Leds::setOutput();
	MODM_LOG_INFO << "\n\nReboot: Ethernet Example" << modm::endl;

	while(true){
		/// ...
		/// we will never reach this ... :(
	}
```

**Comment by rleh** [comment]
@kikass13 The ethernet example is fixed and working perfectly in #760.

```
$ ping -c 1 192.168.178.42
PING 192.168.178.42 (192.168.178.42) 56(84) bytes of data.
64 bytes from 192.168.178.42: icmp_seq=1 ttl=64 time=0.156 ms

--- 192.168.178.42 ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 0.156/0.156/0.156/0.000 ms

$ curl http://192.168.178.42/
<html><body><h1>Hello from your STM32!</h1></body></html>
```

**Comment by kikass13** [comment]
yeah nvm, i applied your fixes and it works now :) thx
